src/deepMerge.ts

```diff
@@ -0,0 +1,112 @@
+import type { MergeDeep } from 'type-fest';
+import { purry } from './purry';
+
+type UnknownRecordOrArray = Record<string, unknown> | Array<unknown>;
+
+function isRecordOrArray(object: unknown): object is UnknownRecordOrArray {
+  return typeof object === 'object' && object !== null;
+}
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param target - The first value to merge. This is the dominant parameter in the merge operation.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.deepMerge(a, b)
+ * @example
+ *    R.deepMerge({ foo: 'bar', x: 1 }, { foo: 'baz', y: 2 }) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_first
+ * @category Object
+ * @pipeable
+ */
+export function deepMerge<
+  Target extends UnknownRecordOrArray,
+  Source extends UnknownRecordOrArray
+>(target: Target, source: Source): MergeDeep<Target, Source>;
```

Comment:
I just noticed, you need to add `{arrayMergeMode: 'spread'}` to the type to cover the concat logic you put in. This also needs to be added in the data-last signature. 
```suggestion
>(target: Target, source: Source): MergeDeep<Target, Source, { arrayMergeMode: "spread" }>;
```